# Profil des applications de l'Apple Store et de Google Play Store

Dans ce projet notre objectif est de trouver le profil des applications profitables pour l'Apple Store et le Google Play Store. Nous travaillons en tant que Data Analyst dans une entreprise réalisant des applications mobiles pour IOS et Android. Notre objectif est de nous appuyer sur la donnée disponible pour orienter les équipes de développeurs dans la construction des applications.

Nous développons uniquement des applications gratuites, ce qui veut dire que notre source princiaple de revenu est la publicité sur notre application. Nous dépendons donc grandement du nombre d'utilisateurs de notre application. L'objectif est donc de déterminer quels types d'application ont le plus du succès.  

# Ouvrir et explorer la donnée

Explorer plus de 4 millions d'applications prendrait beaucoup de temps, nous allons donc analyser un échantillon des applications diponibles. Nous avons choisi d'utiliser 2 bases de données disponibles sur Kaggle. 

 [Une base de données](https://www.kaggle.com/lava18/google-play-store-apps/home) contenant des informations sur environ 10 000 applications de Google Play Store
  [Une base de données](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/home) contenant des informations sur environ 7 000 applications de l'App Store

Commençons par ouvrir et lire les bases de données

In [2]:
from csv import reader
### Google Play Store ###
opened_file = open("googleplaystore.csv")
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

### Apple Store ###
opened_file = open("AppleStore.csv")
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # ajoute une ligne vide après chaque row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

print(android_header)
print('\n')
explore_data(android,0,3,True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


On peut remarquer que la base de données de Google Play Store contient 10841 applications et 13 colonnes. 

Les colonnes utiles sont les suivantes : 'App', 'Category', 'Reviews', 'Installs', 'Type', 'Price', and 'Genres'.

Jetons un coup d'oeil à la base de données de l'IOS

In [4]:
print(ios_header)
print('\n')
explore_data(ios,0,3,True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


On peut remarquer que la base de données de l'IOS contient 7197 applications et 16 colonnes.

Les colonnes utiles sont les suivantes : 'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', and 'prime_genre'.


In [5]:
print(android[10472])
print('\n')
print(android_header)
print('\n')
print(android[0])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


L'erreur que nous venons de corriger était indiquée dans les commentaires de la base de données. 

Nous pouvons remarquer que la ligne 10472 correspond à l'application Life Made WI-Fi Touchscreen Photo Frame, et que le rating de l'application est de 19. La note maximale d'une application étant de 5 nous supprimons cette ligne

In [6]:
print(len(android))
del android[10472]  
print(len(android))

10841
10840


# Retirer les doublons

## Partie une

En explorant la base de données de Google Play store, nous remarquons que des applications ont plusieurs entrées. Comme Instragram qu'on retrouve 4 fois. 

In [7]:
duplicate_apps = []
unique_apps = []

for app in android : 
    name = app[0]
    if name in unique_apps : 
        duplicate_apps.append(name)
    else : 
        unique_apps.append(name)

print('Number of duplicate apps:', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:15])

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


Une fois que nous avons identifié et regroupé les applications en doublons, il nous faut les traiter. Nous pouvons supprimer au hasard les applications pour en garder qu'une mais ce n'est pas l'idéal. En effet, la 4ème colonne indique le nombre d'avis d'utilisateurs. En utilisant cette colonne, nous garderons uniquement les applications ayant le plus d'avis laissés. Cela nous permettra de garder l'application avec les informations les plus recentes. 

Pour cela nous : 

- Construison un dictionnaire avec comme clé, le nom de l'application, et la valeur la plus élevée des commentaires associés à l'application
- Utilisons le dictionnaire pour créer une nouvelle base de données qui aura une ligne par application


## Partie deux

Commençons par construire un dictonnaire.

In [8]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

Nous rappelons que dans un bloc de code précédent, 1 181 applications aparaissaient plus d'une fois. Notre dictionnaire doit être égal à la différence entre la base de donnée initale et 1 181. 

In [9]:
print('Expected length:', len(android) - 1181)
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


In [10]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)

Vérifions que notre nouvelle base de données est bien égale à 9 569.

In [11]:
explore_data(android_clean, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13


# Enlever les applications non-anglaises

## Partie une 

En observant la base de données, on peut observer que certains noms d'applications ne contiennent pas de caractères anglais. 

En voici des exemples

In [12]:

print(ios[813][1])
print(ios[6731][1])

print(android_clean[4412][0])
print(android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜
中国語 AQリスニング
لعبة تقدر تربح DZ


Nous ne sommes pas interessés par ces applications, nous souhaitons donc les enlever. 

Une façon d'enlever ces applications c'est d'identifier toutes celles qui n'utilisent pas de caractères anglais. Pour cela nous utilisons le code ASCII des caractères présents dans le titre. Chaque carractère correspond à un code ASCII, pour les caractères anglais le code ASCII est compris entre 0 et 127.

Nous construisons une fonction pour identifier les code ASCII présents dans le titre.

In [13]:
def is_english(string):
    
    for character in string:
        if ord(character) > 127:
            return False
    
    return True

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
False


Cette fonction semble fonctionner, néanmoins certaines applications utilisent des emojis ou d'autres symmboles qui ne correspondent pas à des caractères anglais. En utilisant la fonction plus haut, nous supprimerons des applications anglaises. 

In [14]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

print(ord('™'))
print(ord('😜'))

False
False
8482
128540


## Partie 2

Nous enlèverons les applications contenant plus de 3 caractères avec un code ASCII supérieur à 127

In [15]:
def is_english(string):
    non_ascii = 0
    
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
    
    if non_ascii > 3:
        return False
    else:
        return True

print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
True


Cette fonction peut encore être améliorée, mais nous nous attardons pas dessus. 

Nous utilisons la fonction définie au dessus sur les 2 bases de données. 

In [16]:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if is_english(name):
        android_english.append(app)
        
for app in ios:
    name = app[1]
    if is_english(name):
        ios_english.append(app)
        
explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

Android : 9614 applications restantes

IOS : 6183 applications restantes 

# Isoler les applications gratuites


In [26]:
freeapps_android = []
freeapps_ios = []

for app in android_english : 
    price =(app[7])
    if price == "0":
        freeapps_android.append(app)

for app in ios_english : 
    price = (app[4])
    if price == "0.0":
        freeapps_ios.append(app)

print(len(freeapps_android))
print(len(freeapps_ios))

8864
3222


Android : 8864 applications restantes 

IOS : 3222 applications restantes

# Applications les plus nombreuses par Genre

## Partie Une

Comme mentionné dans l'introduction, notre objectif est de determiner les types applications qui ont le plus de chance d'attire les utilisateurs.

Pour minimiser les risques, notre stratégie de développement se résume en 3 étapes : 

- Développer une première version de l'application pour Android que nous ajouterons sur Google Play Store 
- Si l'application a du succès, nous développerons une version plus aboutie. 
- Si l'application génère des bénéfices, nous la développerons pour IOS 

Nous allons construire des fonctions pour analyser les "frequency tables" de l'Apple Store et de Google Play Store

Une fonction pour générer une fréquency table et afficher le résultat en pourcentage. Une seconde fonction pour afficher les pourcentages en ordre décroissant. Ce qui facilitera l'analyse. 

In [33]:
def freq_table(dataset,index):
    frequency_table = {}
    for row in dataset : 
        a_datapoint = row[index]
        if a_datapoint in frequency_table : 
            frequency_table[a_datapoint]+=((1*100)/len(dataset))
        else : 
            frequency_table[a_datapoint]=((1*100)/len(dataset))
            
    return frequency_table

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])



## Partie 2

Examinons le résultat pour Android

In [35]:
print(display_table(freeapps_android,1))

FAMILY : 18.907942238266926
GAME : 9.724729241877363
TOOLS : 8.46119133574016
BUSINESS : 4.591606498194979
LIFESTYLE : 3.90342960288811
PRODUCTIVITY : 3.8921480144404565
FINANCE : 3.7003610108303455
MEDICAL : 3.5311371841155417
SPORTS : 3.3957581227436986
PERSONALIZATION : 3.3167870036101235
COMMUNICATION : 3.2378158844765483
HEALTH_AND_FITNESS : 3.079873646209398
PHOTOGRAPHY : 2.944494584837555
NEWS_AND_MAGAZINES : 2.7978339350180583
SOCIAL : 2.6624548736462152
TRAVEL_AND_LOCAL : 2.335288808664261
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.14350180505415
DATING : 1.861462093862813
VIDEO_PLAYERS : 1.7937725631768928
MAPS_AND_NAVIGATION : 1.398916967509025
FOOD_AND_DRINK : 1.2409747292418778
EDUCATION : 1.1620036101083042
ENTERTAINMENT : 0.9589350180505433
LIBRARIES_AND_DEMO : 0.9363718411552363
AUTO_AND_VEHICLES : 0.9250902527075828
HOUSE_AND_HOME : 0.8235559566787015
WEATHER : 0.8009927797833946
EVENTS : 0.7107400722021667
PARENTING : 0.6543321299638993
ART_AND_DESIGN : 0.6

Il y a peu d'applications ludiques sur Android, beaucoup d'applications ont un but pratique (family,tools,business,lifestyle...). Néanmoins en approfondissant les recherches, beaucoup des applications liées au genre "Family" correspondent à des jeux pour enfants. 

Google Play Store comporte plus d'applications "pratiques" qu'IOS (que nous analysons en dessous) 

In [36]:
print(display_table(freeapps_ios,11))

Games : 58.1626319056464
Entertainment : 7.883302296710134
Photo & Video : 4.965859714463075
Education : 3.6623215394165176
Social Networking : 3.2898820608317867
Shopping : 2.6070763500931133
Utilities : 2.5139664804469306
Sports : 2.1415270018621997
Music : 2.048417132216017
Health & Fitness : 2.0173805090006227
Productivity : 1.7380509000620747
Lifestyle : 1.5828677839851035
News : 1.3345747982619496
Travel : 1.2414649286157668
Finance : 1.1173184357541899
Weather : 0.8690254500310364
Food & Drink : 0.8069522036002481
Reference : 0.558659217877095
Business : 0.5276225946617009
Book : 0.4345127250155184
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665
None


Premier constat, plus de la moitié des applications IOS sont des jeux (58.16%). Les applications "Entertainment" représentent environ 8% des applications disponibles sur IOS, 5% pour les applications de photo et vidéo. Seulement 3,66% d'applications "Education", et 3,29% d'applications "Social Networking".

In [37]:
print(display_table(freeapps_android,9))

Tools : 8.449909747292507
Entertainment : 6.069494584837599
Education : 5.34747292418777
Business : 4.591606498194979
Productivity : 3.8921480144404565
Lifestyle : 3.8921480144404565
Finance : 3.7003610108303455
Medical : 3.5311371841155417
Sports : 3.46344765342962
Personalization : 3.3167870036101235
Communication : 3.2378158844765483
Action : 3.1024368231047053
Health & Fitness : 3.079873646209398
Photography : 2.944494584837555
News & Magazines : 2.7978339350180583
Social : 2.6624548736462152
Travel & Local : 2.3240072202166075
Shopping : 2.2450361010830324
Books & Reference : 2.14350180505415
Simulation : 2.041967509025268
Dating : 1.861462093862813
Arcade : 1.8501805054151597
Video Players & Editors : 1.771209386281586
Casual : 1.7599277978339327
Maps & Navigation : 1.398916967509025
Food & Drink : 1.2409747292418778
Puzzle : 1.1281588447653441
Racing : 0.9927797833935037
Role Playing : 0.9363718411552363
Libraries & Demo : 0.9363718411552363
Auto & Vehicles : 0.9250902527075828


Après cette première esquisse d'analyse, on peut s'accorder pour dire qu'alors que l'Apple Store est dominé par les applications ludiques, le Google Play Store est plus équilibré.

Attention, que la moitié des applications disponible sur l'IOS soient des jeux ne veut pas dire que c'est vers une application ludique qu'il faut tendre. D'autres facteurs sont à prendre en compte comme les avis, les téléchargements ou les dates de mise à jour. 

Pour savoir quelles applications ont le plus été téléchargées ( et determiner les marchés saturés ou avec une demande) nous pouvons utiliser le nombre de téléchargement des applications. En faire une moyenne, ou analyser si le marché est saturé par peu d'applications (ex : Transport, oligopole entre Waze et Google Map). 

On peut aussi déterminer, en fonction des moyennes des notes des applications quels sont les genres avec un vide / un besoin utilisateur qui n'a pas été répondu. 

Il reste énormément de points à analyser dans ces bases de données, j'essayerais d'y revenir quand j'aurais plus de temps et de compéténces en développement Python 